src/Faithlife.Utility/EnumerableUtility.cs

```diff
@@ -765,7 +808,122 @@ static IEnumerable<T> DoWhereNotNull(IEnumerable<T?> s)
+			IEnumerable<T2> second,
+			IEnumerable<T3> third,
+			Func<T1, T2, T3, TResult> resultSelector) =>
+			ZipImpl(
+				first ?? throw new ArgumentNullException(nameof(first)),
+				second ?? throw new ArgumentNullException(nameof(second)),
+				third ?? throw new ArgumentNullException(nameof(third)),
+				resultSelector ?? throw new ArgumentNullException(nameof(resultSelector)),
+				UnbalancedZipStrategy.Truncate);
+
+		/// <summary>
+		/// Separates a sequence of tuples into a tuple of sequences.
+		/// </summary>
+		/// <param name="source">The sequence to separate.</param>
+		/// <returns>A tuple of sequences, separating the source sequence's <c>Item1</c> elements from its <c>Item2</c> elements.</returns>
+		[SuppressMessage("StyleCop.CSharp.ReadabilityRules", "SA1414:Tuple types in signatures should have element names", Justification = "By design.")]
+		public static (IEnumerable<T1>, IEnumerable<T2>) Unzip<T1, T2>(this IEnumerable<(T1, T2)> source) =>
+			Unzip(source ?? throw new ArgumentNullException(nameof(source)), x => x.Item1, x => x.Item2);
+
+		/// <summary>
+		/// Separates a sequence of tuples into a tuple of sequences.
+		/// </summary>
+		/// <param name="source">The sequence to separate.</param>
+		/// <returns>A tuple of sequences, separating the source sequence's <c>Item1</c>, <c>Item2</c>, and <c>Item3</c> elements.</returns>
+		[SuppressMessage("StyleCop.CSharp.ReadabilityRules", "SA1414:Tuple types in signatures should have element names", Justification = "By design.")]
+		public static (IEnumerable<T1>, IEnumerable<T2>, IEnumerable<T3>) Unzip<T1, T2, T3>(this IEnumerable<(T1, T2, T3)> source) =>
+			Unzip(source ?? throw new ArgumentNullException(nameof(source)), x => x.Item1, x => x.Item2, x => x.Item3);
+
+		/// <summary>
+		/// Separates a sequence into a tuple of sequences.
+		/// </summary>
+		/// <param name="source">The sequence to separate.</param>
+		/// <param name="firstSelector">A function for selecting the first return sequence's elements from the source sequence's elements.</param>
+		/// <param name="secondSelector">A function for selecting the second return sequence's elements from the source sequence's elements.</param>
+		/// <returns>A tuple of sequences, separating the source sequence according to the selector arguments.</returns>
+		[SuppressMessage("StyleCop.CSharp.ReadabilityRules", "SA1414:Tuple types in signatures should have element names", Justification = "By design.")]
+		public static (IEnumerable<T1>, IEnumerable<T2>) Unzip<TSource, T1, T2>(
+			this IEnumerable<TSource> source,
+			Func<TSource, T1> firstSelector,
+			Func<TSource, T2> secondSelector)
+		{
+			var enumeratedSource = source?.ToArray() ?? throw new ArgumentNullException(nameof(source));
+			return (
+				enumeratedSource.Select(firstSelector ?? throw new ArgumentNullException(nameof(firstSelector))),
+				enumeratedSource.Select(secondSelector ?? throw new ArgumentNullException(nameof(secondSelector)))
```

Comment:
Should I remove those `?? throw new ArgumentNullException`s since `ToArray` and `Select` already throw the same thing for `null` arguments? Or is it better to add the throw at this level, too?